components/masking_functions/src/masking_functions/registration_routines.cpp

```diff
@@ -0,0 +1,1309 @@
+// gen_dictionary(string)
+//
+// This function returns a random term from the dictionary whose name
+// specified via the first argument.
+class gen_dictionary_impl {
+ public:
+  gen_dictionary_impl(mysqlpp::udf_context &ctx) {
+    if (ctx.get_number_of_args() != 1)
+      throw std::invalid_argument{
+          "Wrong argument list: gen_dictionary(string)"};
+
+    ctx.mark_result_nullable(true);
+    ctx.mark_result_const(false);
+
+    // arg0 - dictionary
+    ctx.mark_arg_nullable(0, false);
+    ctx.set_arg_type(0, STRING_RESULT);
+
+    mysqlpp::udf_context_charset_extension charset_ext{
+        mysql_service_mysql_udf_metadata};
+    charset_ext.set_return_value_collation(
+        ctx, masking_functions::charset_string::default_collation_name);
+  }
+
+  mysqlpp::udf_result_t<STRING_RESULT> calculate(
+      const mysqlpp::udf_context &ctx) {
+    const auto cs_dictionary = make_charset_string_from_arg(ctx, 0);
+
+    masking_functions::sql_context sql_ctx{global_command_services::instance()};
+
+    auto query =
+        global_query_builder::instance().select_random_term_for_dictionary(
+            cs_dictionary);
+    auto sresult = sql_ctx.query_single_value(query);
+
+    if (sresult && sresult->size() > 0) {
+      return *sresult;
+    } else {
+      return std::nullopt;
+    }
+  }
+};
+
+//
+// masking_dictionary_remove(string)
+//
+// This function removes the dictionary, whose name is specified via the first
+// argument, and all of its terms from the dictionary registry.
+class masking_dictionary_remove_impl {
+ public:
+  masking_dictionary_remove_impl(mysqlpp::udf_context &ctx) {
+    if (!have_masking_admin_privilege()) {
+      throw std::invalid_argument{
+          "Function requires " +
+          std::string(masking_dictionaries_privilege_name) + " privilege"};
+    }
+
+    if (ctx.get_number_of_args() != 1)
+      throw std::invalid_argument{
+          "Wrong argument list: masking_dictionary_remove(string)"};
+
+    ctx.mark_result_nullable(true);
+    // Calling this UDF two or more times has exactly the same effect as just
+    // calling it once. So, we mark the result as 'const' here so that the
+    // optimizer could use this info to eliminate unnecessary calls.
+    ctx.mark_result_const(true);
+
+    // arg0 - dictionary
+    ctx.mark_arg_nullable(0, false);
+    ctx.set_arg_type(0, STRING_RESULT);
+
+    mysqlpp::udf_context_charset_extension charset_ext{
+        mysql_service_mysql_udf_metadata};
+    charset_ext.set_return_value_collation(
+        ctx, masking_functions::charset_string::default_collation_name);
+  }
+
+  mysqlpp::udf_result_t<STRING_RESULT> calculate(
+      const mysqlpp::udf_context &ctx) {
+    const auto cs_dictionary = make_charset_string_from_arg(ctx, 0);
+
+    masking_functions::sql_context sql_ctx{global_command_services::instance()};
+
+    auto query =
+        global_query_builder::instance().delete_for_dictionary(cs_dictionary);
+    if (!sql_ctx.execute(query)) {
```

Comment:
There is one more issue which I have missed during previous phase of my review.

Dictionary functions which modify dictionary table won't work as one would expect in case if user database has GLOBAL autocommit variable set to 0 (or in other cases when server doesn't commit statements by default). The problem is that virtual connection used for Sql_context will inherit its autocommit setting from global value and thus INSERT and DELETE statements executed by dictionary UDFs won't commit, and will be rolled back instead when virtual connection closes.
I have written a small test case which demonstrates this:

INSTALL COMPONENT 'file://component_masking_functions';
CREATE TABLE mysql.masking_dictionaries(
Dictionary VARCHAR(256) NOT NULL,
Term VARCHAR(256) NOT NULL,
UNIQUE INDEX dictionary_term_idx (Dictionary, Term)
) ENGINE = InnoDB DEFAULT CHARSET=utf8mb4;
SELECT CURRENT_USER();
// CURRENT_USER()
// root@localhost
GRANT MASKING_DICTIONARIES_ADMIN ON *.* TO root@localhost;
SELECT masking_dictionary_term_add('single_dict', 'entry');
// masking_dictionary_term_add('single_dict', 'entry')
// 1
SET @@global.autocommit = 0;
SELECT masking_dictionary_term_add('single_dict', 'another_entry');
// masking_dictionary_term_add('single_dict', 'another_entry')
// 1
SELECT * FROM mysql.masking_dictionaries;
// Dictionary	Term
// single_dict	entry
COMMIT;
SELECT * FROM mysql.masking_dictionaries;
// Dictionary	Term
// single_dict	entry
SET @@global.autocommit = 1;
SELECT masking_dictionary_term_add('single_dict', 'yet_another_entry');
// masking_dictionary_term_add('single_dict', 'yet_another_entry')
// 1
SELECT * FROM mysql.masking_dictionaries;
// Dictionary	Term
// single_dict	entry
// single_dict	yet_another_entry

Notice total absence of 'another_entry' term in the table.
I think we either need to turn on autocommit mode when virtual connection is created in SQL-context, or we need to commit INSERT and DELETE statements we run from dictionary functions explicitly.

What do you think?

Reply:
Good point. I will make this scenario an MTR test case and will extend `sql_context` constructor with `set auto_commit = ON`

Reply:
Great! Thank you!